### PR TITLE
fix: allow ranks to report independent early exits [DET-6792, DET-6835]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change modifies the searcher logic to allow multiple invalid HP exits per trial. Exits after the first are ignored. This is useful for multi-rank trials to invalid HP without coordination. The change in behavior in this change is that now, rather than the experiment failing violently on the receipt of the 2nd invalid HP, nothing will happen. Nothing includes no kill, cancel, etc. The master will just wait for the trial to exit on its own, so it is a bit susceptible to hangs.

## Test Plan
- [ ] run invalid hp dtrain, see less failures

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: master/internal/trial.go [hunk 1]
```diff
@@ -235,6 +242,25 @@ func (t *trial) maybeAllocateTask(ctx *actor.Context) error {
 	return nil
 }
 
+const (
+	// InvalidHPKillDelay the delay before we forcibly kill a trial that said it had an invalid HP.
+	InvalidHPKillDelay = 10 * time.Second
```
Comment (liamcli) [inline master/internal/trial.go:247]: can we lower this to 3 seconds.  that matches the grace period we give the launched processes in the harness

Comment (stoksc) [inline master/internal/trial.go:247]:
```suggestion
	InvalidHPKillDelay = 3 * time.Second
```

File: master/internal/experiment.go [hunk 2]
```diff
@@ -54,11 +54,19 @@ type (
...
+
+	// trialClosed is used to replay closes missed when the master dies between when a trial closing
+	// in its actor.PostStop and when the experiment snapshots the trial closed.
 	trialClosed struct {
 		requestID model.RequestID
 	}
+
+	// userInitiatedEarlyExit is a user-injected message, provided through the early exit API. It
+	// _should_ indicate the user is exiting, but in the event they don't, we will clean them up.
+	userInitiatedEarlyExit struct {
```
Comment (liamcli) [inline master/internal/experiment.go:66]: Do we still need type `trialReportEarlyExit` above and related code in experiment?

Comment (stoksc) [reply master/internal/experiment.go:66]: yeah. the flow here now is a `userInitiatedEarlyExit` is sent to the trial and the trial tells the experiment `trialReportEarlyExit` once it is done and closed.

**Review by liamcli — APPROVED** [review]
i tested locally and lgtm

**Pull request merged**